internal/controller/environments/environments.go

```diff
@@ -170,6 +181,28 @@ func SetupReconcilerWithManager(
 		Complete(e)
 }
 
+func indexEnvsByApp(obj client.Object) []string {
+	env := obj.(*api.Environment) // nolint: forcetypeassert
+	if env.Spec.HealthChecks == nil {
+		return nil
+	}
+	apps := make([]string, len(env.Spec.HealthChecks.ArgoCDAppChecks))
+	for i, appCheck := range env.Spec.HealthChecks.ArgoCDAppChecks {
+		apps[i] =
+			fmt.Sprintf("%s:%s", appCheck.AppNamespace, appCheck.AppName)
+	}
+	return apps
+}
+
+func indexOutstandingPromotionsByEnvironment(obj client.Object) []string {
+	promo := obj.(*api.Promotion) // nolint: forcetypeassert
+	switch promo.Status.Phase {
+	case api.PromotionPhaseComplete, api.PromotionPhaseFailed:
+		return nil
+	}
+	return []string{promo.Spec.Environment}
+}
+
```

Comment:
This indexes a promotion to an Environment _only_ if its in a non-terminal state.